### PR TITLE
Initial draft of capabilities support in OdcmModel

### PR DESCRIPTION
**Pull request by Sridhar-MS**
- This commit is just for reviewing. It will not be pushed into the repo.
- Created a new class 'OdcmProjection' which captures the OdcmType and OdcmCapabilities.
- Created a new property called 'Projection' (type - OdcmProjection) for OdcmProperty class.
- Based on Capabilities annotations populate the Projection property of OdcmProperty (only for entity sets and navigation properties).
- OdcmProjection.Capabilities will be null if there are no capabilities annotations in the edmx. Writers can assume that if Capabilities is null then emit the default code else emit code which projects those capabilities.
- No unit tests yet. Will add them once the design is finalized.
- OdcmProperty.Type is now redundant and needs to be removed. TBD.

### CODE REVIEW

File: src/Core/Vipr.Core/CodeModel/OdcmProjection.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Vipr.Core.CodeModel
+{
+    public class OdcmProjection
+    {
+        public OdcmType Type { get; set; }
+
+        public OdcmCapabilities Capabilities { get; set; }
+
+        public OdcmCapabilities EnsureCapabilities()
```
Comment (MrTomWhite) [inline src/Core/Vipr.Core/CodeModel/OdcmProjection.cs]: Why do we need this? What's the call pattern for `EnsureCapabilties()`?

File: src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+using Vipr.Core;
+using Vipr.Core.CodeModel;
+
+namespace ODataReader.v4
+{
+    internal static class ODataCapabilitiesReader
+    {
+        private static IEdmModel s_capabilitiesModel;
+        private static readonly IEdmValueTerm s_insertRestrictionsTerm;
+        private static readonly IEdmValueTerm s_updateRestrictionsTerm;
```
Comment (MrTomWhite) [inline src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs]: This is going to turn into a rehash of CapabilitiesHelper from the OData.net folks. We already had some code that parses annotations, why do we need more?

Comment (MrTomWhite) [inline src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs]: +This boiler plate all follows from the Edm type of the annotation itself.

File: src/Readers/ODataReader.v4/OdcmReader.cs [hunk 3]
```diff
@@ -421,6 +431,11 @@ private void WriteProperty(OdcmClass odcmClass, IEdmSingleton singleton)
                 {
                     Class = odcmClass,
                     Type = ResolveType(singleton.EntityType().Name, singleton.EntityType().Namespace),
+                    Projection = new OdcmProjection()
+                    {
+                        Type = ResolveType(singleton.EntityType().Name, singleton.EntityType().Namespace),
+                        Capabilities = null
```
Comment (MrTomWhite) [inline src/Readers/ODataReader.v4/OdcmReader.cs]: Why use a null here instead of a default capabilities object?

File: src/Core/Vipr.Core/CodeModel/OdcmCapabilities.cs [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Vipr.Core.CodeModel
+{
+    public class OdcmCapabilities
+    {
+        public bool Insertable { get; set; }
```
Comment (piotrpMSFT) [inline src/Core/Vipr.Core/CodeModel/OdcmCapabilities.cs]: This was originally discussed as a collection of OdcmCapability with writers intended to iterate over the collection. Why the change of direction? Did that design prove too cumbersome?

File: src/Core/Vipr.Core/CodeModel/OdcmProjection.cs (hunk 1 above)
Comment (piotrpMSFT) [inline src/Core/Vipr.Core/CodeModel/OdcmProjection.cs]: Agreed. Can't we capture this in the constructor or, if laziness is the intent, in the getter?

File: src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+                OdcmProperty navProperty;
+                if (!(odcmProperty.Type as OdcmClass).TryFindProperty(propertyPath, out navProperty))
+                {
+                    throw new InvalidOperationException();
+                }
+                navProperty.Projection.EnsureCapabilities().Expandable = false;
+            }
+        }
+
+        private static void GetBooleanAndPathCollection(IEdmValueAnnotation annotation, string booleanPropertyName, string pathsPropertyName, out bool boolean, out IEnumerable<string> paths)
```
Comment (piotrpMSFT) [inline src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs]: Can we use the return value instead of an out parameter? Out parameters are generally used in TrySomething(out result) scenarios.

File: src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+                OdcmProperty navProperty;
+                if (!(odcmProperty.Type as OdcmClass).TryFindProperty(propertyPath, out navProperty))
+                {
+                    throw new InvalidOperationException();
+                }
+                navProperty.Projection.EnsureCapabilities().Deleteable = false;
+            }
+        }
+
+        private static void GetExpandRestrictions(OdcmProperty odcmProperty, IEdmModel model, IEdmEntitySet entitySet)
```
Comment (piotrpMSFT) [inline src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs]: The name here is confusing because the method does not return a value. Consider modeling this as an extension method on odcmProperty, showing who is being acted upon, and rename it to something like SetNavigationProperty*Restrictions

File: src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+                if (!(odcmProperty.Type as OdcmClass).TryFindProperty(propertyPath, out navProperty))
+                {
+                    throw new InvalidOperationException();
+                }
+                navProperty.Projection.EnsureCapabilities().Updateable = false;
+            }
+        }
+
+        private static void GetDeleteRestrictions(OdcmProperty odcmProperty, IEdmModel model, IEdmEntitySet entitySet)
+        {
```
Comment (piotrpMSFT) [inline src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs]: The name here is confusing because the method does not return a value. Consider modeling this as an extension method on odcmProperty, showing who is being acted upon, and rename it to something like SetNavigationProperty*Restrictions

File: src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+                if (!(odcmProperty.Type as OdcmClass).TryFindProperty(propertyPath, out navProperty))
+                {
+                    throw new InvalidOperationException();
+                }
+                navProperty.Projection.EnsureCapabilities().Insertable = false;
+            }
+        }
+
+        private static void GetUpdateRestrictions(OdcmProperty odcmProperty, IEdmModel model, IEdmEntitySet entitySet)
+        {
```
Comment (piotrpMSFT) [inline src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs]: The name here is confusing because the method does not return a value. Consider modeling this as an extension method on odcmProperty, showing who is being acted upon, and rename it to something like SetNavigationProperty*Restrictions

File: src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+            GetInsertRestrictions(odcmProperty, edmModel, entitySet);
+            GetUpdateRestrictions(odcmProperty, edmModel, entitySet);
+            GetDeleteRestrictions(odcmProperty, edmModel, entitySet);
+            GetExpandRestrictions(odcmProperty, edmModel, entitySet);
+        }
+
+        private static void GetInsertRestrictions(OdcmProperty odcmProperty, IEdmModel model, IEdmEntitySet entitySet)
+        {
+            IEdmValueAnnotation insertAnnotation = FindVocabularyAnnotation(model, entitySet, s_insertRestrictionsTerm);
```
Comment (piotrpMSFT) [inline src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs]: The name here is confusing because the method does not return a value. Consider modeling this as an extension method on odcmProperty, showing who is being acted upon, and rename it to something like SetNavigationProperty*Restrictions

File: src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+                }
+            }
+
+            s_insertRestrictionsTerm = s_capabilitiesModel.FindDeclaredValueTerm(CapabilitiesInsertRestrictions);
+            s_updateRestrictionsTerm = s_capabilitiesModel.FindDeclaredValueTerm(CapabilitiesUpdateRestrictions);
+            s_deleteRestrictionsTerm = s_capabilitiesModel.FindDeclaredValueTerm(CapabilitiesDeleteRestrictions);
+            s_expandRestrictionsTerm = s_capabilitiesModel.FindDeclaredValueTerm(CapabilitiesExpandRestrictions);
+        }
+
+        internal static void GetCapabilitiesForEntitySet(OdcmProperty odcmProperty, IEdmModel edmModel, IEdmEntitySet entitySet)
```
Comment (piotrpMSFT) [inline src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs]: The name here is confusing because the method does not return a value. Consider modeling this as an extension method on odcmProperty, showing who is being acted upon, and rename it to something like SetNavigationProperty*Restrictions

File: src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs (hunk 9 above)
Comment (piotrpMSFT) [inline src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs]: It seems like each of these private methods does approximately the same action but with different constants for reading annotations. Consider refactoring the common logic into a generalized method and calling that method with specific constants.

File: src/Readers/ODataReader.v4/OdcmReader.cs [hunk 11]
```diff
@@ -406,6 +411,11 @@ private void WriteProperty(OdcmClass odcmClass, IEdmEntitySet entitySet)
                 {
                     Class = odcmClass,
                     Type = ResolveType(entitySet.EntityType().Name, entitySet.EntityType().Namespace),
+                    Projection = new OdcmProjection()
```
Comment (piotrpMSFT) [inline src/Readers/ODataReader.v4/OdcmReader.cs]: Shouldn't we be de-duping Projections here? I think we'd be safe to assume that so long as two Properties require the same capabilities then it's safe to re-use an existing projection.

File: src/Core/Vipr.Core/CodeModel/OdcmCapabilities.cs (hunk 4 above)
Comment (MrTomWhite) [inline src/Core/Vipr.Core/CodeModel/OdcmCapabilities.cs]: I don't think the OdcmCapabilties representation is mutually excusive with the originally discussed model. (You can create the original model -> parse the original model to get a OdcmCapabiltiies instance)
The current implementation reimplements the original parsing specified to items in the capabilities vocabulary and then populates the OdcmCapabilities datastructure. I don't think that's the right approach. :(

File: src/Core/Vipr.Core/CodeModel/OdcmProjection.cs (hunk 1 above)
Comment (Sridhar-MS) [inline src/Core/Vipr.Core/CodeModel/OdcmProjection.cs]: In the new commit, OdcmProjection has a list of OdcmCapability.

File: src/Readers/ODataReader.v4/OdcmReader.cs (hunk 3 above)
Comment (Sridhar-MS) [inline src/Readers/ODataReader.v4/OdcmReader.cs]: Changed this to get a List of default OdcmCapability for an OdcmProjection.

File: src/Core/Vipr.Core/CodeModel/OdcmCapabilities.cs (hunk 4 above)
Comment (Sridhar-MS) [inline src/Core/Vipr.Core/CodeModel/OdcmCapabilities.cs]: Changed this to use a collection of OdcmCapability. Now Writers can iterate over this collection to get a list of capability. it also make adding new capabilities support a bit easy.

File: src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs (hunk 5 above)
Comment (Sridhar-MS) [inline src/Readers/ODataReader.v4/ODataCapabilitiesReader.cs]: Refactored this method into 2 different 'Get' methods.

File: src/Core/Vipr.Core/CodeModel/OdcmCapabilities.cs (hunk 4 above)
Comment (Sridhar-MS) [inline src/Core/Vipr.Core/CodeModel/OdcmCapabilities.cs]: @MrTomWhite I feel the original model adds an extra layer in converting the raw annotations into OdcmCapability. Also in this case the raw annotation (like insertrestriction) properties do not map directly to OdcmCapbility class. 

Also OdataCapabiltiesReader right now dispatches the parsing logic to a list of parsers which can parse the annotations any way they want.

Comment (MrTomWhite) [inline src/Core/Vipr.Core/CodeModel/OdcmCapabilities.cs]: >"I feel the original model adds an extra layer in converting the raw annotations into OdcmCapability." 

We don't really need an extra level of abstraction here do we? This "Extra level" is going to produce a collection of capability objects which will have roughly the same set of call patterns from the other proposal; it doesn't feel like it makes our lives easier and adds some complexity. 

>"Also OdataCapabiltiesReader right now dispatches the parsing logic to a list of parsers which can parse the annotations any way they want. " 

I don't really agree that this is nescessary. There's not going to be custom parsing semantics for annotations that are already defined in some schema somewhere. The "custom parsing" is just defined as being able to interpret the EDM model, which is parsable fully from knowledge of the types. 

imho the "list of parsers" is just extra code we dont' need + extra complexity.

Comment (Sridhar-MS) [inline src/Core/Vipr.Core/CodeModel/OdcmCapabilities.cs]: >" There's not going to be custom parsing semantics for annotations that are already defined in some schema somewhere."

This is not true. For example take the case UpdateRestriction annotation. We do not directly parse and convert this annotation into strongly typed CLR object using the generic parser. That is not very useful in how we want to represent OdcmCapability for OdcmProperty. Instead we do custom parsing to represent this annotation in the OdcmModel.
